cpp/modmesh/buffer/BufferExpander.hpp

```diff
@@ -0,0 +1,173 @@
+ * - Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ * - Neither the name of the copyright holder nor the names of its contributors
+ *   may be used to endorse or promote products derived from this software
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <modmesh/base.hpp>
+#include <modmesh/buffer/ConcreteBuffer.hpp>
+
+#include <stdexcept>
+#include <memory>
+#include <algorithm>
+#include <sstream>
+
+namespace modmesh
+{
+
+/**
+ * Untyped and growing memory buffer for contiguous data storage.  The internal
+ * expandable memory buffer cannot be used externally.
+ */
+class BufferExpander
+    : public std::enable_shared_from_this<BufferExpander>
+{
+
+private:
+
+    struct ctor_passkey
+    {
+    };
+
+public:
+
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+
+    template <typename... Args>
+    static std::shared_ptr<BufferExpander> construct(Args &&... args)
+    {
+        return std::make_shared<BufferExpander>(std::forward<Args>(args)..., ctor_passkey());
+    }
+
+    std::shared_ptr<BufferExpander> clone()
+    {
+        return BufferExpander::construct(copy_concrete(), /*clone*/ false);
+    }
+
+    BufferExpander(std::shared_ptr<ConcreteBuffer> const & buf, bool clone, ctor_passkey const &)
+        : m_concrete_buffer(clone ? buf->clone() : buf)
+        , m_begin(m_concrete_buffer->data())
+        , m_end(m_begin + size())
+        , m_end_cap(m_begin + m_concrete_buffer->size())
+    {
+    }
+
+    BufferExpander(size_type nbyte, ctor_passkey const &)
+    {
+        expand(nbyte);
+    }
+
+    BufferExpander(ctor_passkey const &) {}
+
+    BufferExpander() = delete;
+    BufferExpander(BufferExpander const &) = delete;
+    BufferExpander(BufferExpander &&) = delete;
+    BufferExpander & operator=(BufferExpander const &) = delete;
+    BufferExpander & operator=(BufferExpander &&) = delete;
+    ~BufferExpander() = default;
+
+    size_type size() const noexcept
```

Comment:
in ConcreteBuffer`, we also have `nbytes()` method, not sure what is the original intention, but maybe we can also add it.

Reply:
It's a good idea!  There are other helpers available in `ConcreteBuffer` but not in `BufferExpander`.  Some of them may be shared between the two types.

After the PR is merged, we may consider two things.  One is to refactor `ConcreteBuffer` and `BufferExpander` by providing mixins to collect common (and simple) code together.  The other is to supplement missing helpers.

I guess it's good for the two things to be done separately for clarity.

Reply:
Issues #305 and #306 are created.